### PR TITLE
gh-111058: Change coro.cr_frame/gen.gi_frame to be None for a closed coroutine/generator.

### LINKED ISSUE #111058: Python 3.12 inspecting a coroutine using `getcoroutinestate` returns the wrong state.
# Bug report

### Bug description:

Prior to Python 3.12, we can close a just created coroutine, thus, the following test passes:

```python
from inspect import getcoroutinestate, CORO_CLOSED


async def do_nothing():
    pass


def test_immediate_close():
    coro = do_nothing()
    coro.close()
    assert getcoroutinestate(coro) == CORO_CLOSED
```

But in Python 3.12, the assertion fails and `getcoroutinestate(coro)` still is `CORO_CREATED` after the `coro.close()`. You can confirm it from https://github.com/gottadiveintopython/py312/actions/runs/6571303025/job/17850194150.

I don't know if this is an intensional change or not, but I just wanted to tell you.

### CPython versions tested on:

3.12

### Operating systems tested on:

Linux

### Linked PRs
* gh-112428
* gh-112589

### PR DESCRIPTION
**Pull request by iritkatriel**
Fixes #111058.

* Issue: gh-111058

### CODE REVIEW

File: Lib/test/test_inspect/test_inspect.py [hunk 1]
```diff
@@ -2493,6 +2493,10 @@ def test_closed_after_immediate_exception(self):
             self.coroutine.throw(RuntimeError)
         self.assertEqual(self._coroutinestate(), inspect.CORO_CLOSED)
 
+    def test_closed_after_close(self):
+        self.coroutine.close()
+        self.assertEqual(self._coroutinestate(), inspect.CORO_CLOSED)
```
Comment (Eclips4) [inline Lib/test/test_inspect/test_inspect.py:2502]: Since this issue applies to generators as well, should we add a test for them?

**Review by Eclips4 — APPROVED** [review]
LGTM. Maybe we need more general `NEWS` entry because the issue applies not only to coroutines but also to generators?

File: Misc/NEWS.d/next/Core and Builtins/2023-11-26-21-30-11.gh-issue-111058.q4DqDY.rst [hunk 2]
```diff
@@ -0,0 +1,3 @@
+Change coro.cr_frame to return ``None`` after the coroutine has been closed.
+This fixes a bug where :func:`~inspect.getcoroutinestate` returns the wrong
+state for a closed coroutine.
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core and Builtins/2023-11-26-21-30-11.gh-issue-111058.q4DqDY.rst]:
```suggestion
Change coro.cr_frame/gen.gi_frame to return ``None`` after the coroutine/generator has been closed.
This fixes a bug where :func:`~inspect.getcoroutinestate` and :fund:`~inspect.getgeneratorstate`
return the wrong state for a closed coroutine/generator.
```

File: Misc/NEWS.d/next/Core and Builtins/2023-11-26-21-30-11.gh-issue-111058.q4DqDY.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Change coro.cr_frame/gen.gi_frame to return ``None`` after the coroutine/generator has been closed.
+This fixes a bug where :func:`~inspect.getcoroutinestate` and :fund:`~inspect.getgeneratorstate`
```
Comment (iritkatriel) [inline Misc/NEWS.d/next/Core and Builtins/2023-11-26-21-30-11.gh-issue-111058.q4DqDY.rst]:
```suggestion
This fixes a bug where :func:`~inspect.getcoroutinestate` and :func:`~inspect.getgeneratorstate`
```

**Review by markshannon — APPROVED** [review]
Looks good. Thanks for fixing this.

**Pull request merged**